crml/transaction-payment/src/lib.rs

```diff
@@ -243,48 +241,46 @@ where
 		info: Self::DispatchInfo,
 		len: usize,
 	) -> TransactionValidity {
-		// A cautionary measure to ensure the storage of previous transactions wouldn't extend beyond them.
-		unhashed::kill(&GAS_FEE_EXCHANGE_KEY);
-
-		let tip = self.tip;
-
-		// pay any fees.
-		let fee = Self::compute_fee(len as u32, info, tip);
+		let fee = Self::compute_fee(len as u32, info, self.tip);
 
-		let mut exchange_op = self.fee_exchange.clone();
-
-		// Only mess with balances if fee is not zero.
+		// Only mess with balances if the fee is not zero.
 		if !fee.is_zero() {
-			if let Some(exchange) = exchange_op {
-				let sold_amount = T::BuyFeeAsset::buy_fee_asset(who, fee, &exchange)
-					.map_err(|_| TransactionValidityError::from(InvalidTransaction::Payment))?;
-				let new_fee_exchange_max = exchange
-					.get_balance()
-					.checked_sub(&sold_amount)
-					.ok_or(TransactionValidityError::from(InvalidTransaction::Payment))?;
-				exchange_op = Some(<TransactionFeeExchange<T>>::new_v1(
-					exchange.get_asset_id(),
-					new_fee_exchange_max,
-				));
+			if let Some(exchange) = &self.fee_exchange {
+				// Buy the CENNZnet fee currency paying with the user's nominated fee currency
+				let fees_purchased = T::BuyFeeAsset::buy_fee_asset(who, fee, &exchange)
+					.map_err(|_| TransactionValidityError::Invalid(InvalidTransaction::Payment))?;
+
+				// Temporarily store the `FeeExchange` info so that it can be used later by the-
+				// custom `GasHandler` impl see `runtime/src/impls.rs`
+				// This is a hack!
+				storage::unhashed::put(
+					&GAS_FEE_EXCHANGE_KEY,
+					&FeeExchange::new_v1(
+						exchange.asset_id(),
+						exchange.max_payment().checked_sub(&fees_purchased).unwrap_or(0.into()),
+					),
+				);
+			} else {
+				// Pre-caution to ensure `FeeExchange` data is cleared
+				storage::unhashed::kill(&GAS_FEE_EXCHANGE_KEY);
```

Comment:
Because it was a cautionary measure to address possible past transaction's mess, the condition for doing that couldn't be subject to anything from the current transaction. Thus it was placed at the top with no conditions. We do that because the time for the previous transactions has come to an end and it's a new one.